flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/NettyMessageClientDecoderDelegateTest.java

```diff
@@ -0,0 +1,377 @@
+			readInputAndVerify(channel, splitBuffers, messages);
+		} finally {
+			if (serializedBuffers != null) {
+				releaseBuffers(serializedBuffers);
+			}
+
+			if (mergedBuffer != null) {
+				mergedBuffer.release();
+			}
+		}
+	}
+
+	/**
+	 * Verifies that the client side decoder works well for empty buffers. Empty buffers should not
+	 * consume data buffers of the input channels.
+	 */
+	@Test
+	public void testDownstreamMessageDecodeWithEmptyBuffers() throws Exception {
+		// 4 buffers required for running 2 rounds and 2 buffers each round.
+		EmbeddedChannel channel = new EmbeddedChannel(
+			new NettyMessageClientDecoderDelegate(createPartitionRequestClientHandler(4)));
+		Buffer event = createDataBuffer(32);
+		event.tagAsEvent();
+
+		NettyMessage[] messages = new NettyMessage[]{
+			new NettyMessage.BufferResponse(createDataBuffer(128), 0, NORMAL_INPUT_CHANNEL_ID, 4),
+			new NettyMessage.BufferResponse(createDataBuffer(0), 1, NORMAL_INPUT_CHANNEL_ID, 3),
+			new NettyMessage.BufferResponse(event, 2, NORMAL_INPUT_CHANNEL_ID, 4),
+			new NettyMessage.ErrorResponse(new RuntimeException("test"), NORMAL_INPUT_CHANNEL_ID),
+			new NettyMessage.BufferResponse(createDataBuffer(56), 3, NORMAL_INPUT_CHANNEL_ID, 4)
+		};
+
+		ByteBuf[] serializedBuffers = null;
+		ByteBuf mergedBuffer = null;
+
+		try {
+			serializedBuffers = serializeMessages(messages);
+			int[] sizes = getBufferSizes(serializedBuffers);
+			mergedBuffer = mergeBuffers(serializedBuffers);
+
+			ByteBuf[] splitBuffers = partitionBuffer(mergedBuffer, new int[]{
+				sizes[0] / 3,
+				sizes[0] + sizes[1] + sizes[2] / 3 * 2,
+				sizes[0] + sizes[1] + sizes[2] + sizes[3] / 3,
+				sizes[0] + sizes[1] + sizes[2] + sizes[3] + sizes[4]
+			});
+			readInputAndVerify(channel, splitBuffers, messages);
+
+			splitBuffers = partitionBuffer(mergedBuffer, new int[]{
+				sizes[0] / 3,
+				sizes[0] + sizes[1] / 3,
+				sizes[0] + sizes[1] + sizes[2] / 3,
+				sizes[0] + sizes[1] + sizes[2] + sizes[3],
+				sizes[0] + sizes[1] + sizes[2] + sizes[3] + sizes[4] / 3 * 2
+			});
+			readInputAndVerify(channel, splitBuffers, messages);
+		} finally {
+			if (serializedBuffers != null) {
+				releaseBuffers(serializedBuffers);
+			}
+
+			if (mergedBuffer != null) {
+				mergedBuffer.release();
+			}
+		}
+	}
+
+	/**
+	 * Verifies that NettyMessageDecoder works well with buffers sent to a released channel.
+	 * For such a channel, no Buffer is available to receive the data buffer in the message,
+	 * and the data buffer part should be discarded before reading the next message.
+	 */
+	@Test
+	public void testDownstreamMessageDecodeWithReleasedInputChannel() throws Exception {
+		// 6 buffers required for running 2 rounds and 3 buffers each round.
+		EmbeddedChannel channel = new EmbeddedChannel(
+			new NettyMessageClientDecoderDelegate(createPartitionRequestClientHandler(6)));
+
+		Buffer event = createDataBuffer(32);
+		event.tagAsEvent();
+
+		NettyMessage[] messages = new NettyMessage[]{
```

Comment:
I see many common codes in these tests which could be extracted into a separate method. E.g. the netty message generation and finally region.